presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java

```diff
@@ -0,0 +1,436 @@
+            Set<Symbol> masks = ImmutableSet.copyOf(node.getMasks().values());
+            if (masks.size() != 1 || node.getMasks().size() == node.getAggregations().size()) {
+                return context.defaultRewrite(node, Optional.empty());
+            }
+
+            AggregateInfo aggregateInfo = new AggregateInfo(node.getGroupBy(),
+                    Iterables.getOnlyElement(masks),
+                    node.getAggregations(),
+                    node.getFunctions());
+
+            PlanNode source = context.rewrite(node.getSource(), Optional.of(aggregateInfo));
+
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregations = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Signature> functions = ImmutableMap.builder();
+            for (Map.Entry<Symbol, FunctionCall> entry : node.getAggregations().entrySet()) {
+                FunctionCall functionCall = entry.getValue();
+                if (entry.getValue().isDistinct()) {
+                    aggregations.put(entry.getKey(), new FunctionCall(functionCall.getName(),
+                            functionCall.getWindow(),
+                            false,
+                            ImmutableList.of(Iterables.getOnlyElement(aggregateInfo.getNewDistinctAggregateSymbols().values()).toSymbolReference())));
+                    functions.put(entry.getKey(), node.getFunctions().get(entry.getKey()));
+                }
+                else {
+                    // Aggregations on non-distinct are already done by new node, just extract the non-null value
+                    Symbol argument = aggregateInfo.getNewNonDistinctAggregateSymbols().get(entry.getKey());
+                    QualifiedName functionName = QualifiedName.of("arbitrary");
+                    aggregations.put(entry.getKey(), new FunctionCall(functionName,
+                            functionCall.getWindow(),
+                            false,
+                            ImmutableList.of(argument.toSymbolReference())));
+                    functions.put(entry.getKey(),
+                            metadata.getFunctionRegistry().resolveFunction(functionName, ImmutableList.of(symbolAllocator.getTypes().get(argument).getTypeSignature()), false));
+                }
+            }
+
+            return new AggregationNode(idAllocator.getNextId(),
+                    source,
+                    node.getGroupBy(),
+                    aggregations.build(),
+                    functions.build(),
+                    Collections.emptyMap(),
+                    node.getGroupingSets(),
+                    node.getStep(),
+                    node.getSampleWeight(),
+                    node.getConfidence(),
+                    Optional.<Symbol>empty());
+        }
+
+        @Override
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<AggregateInfo>> context)
```

Comment:
If you code extract parts of this method to other methods - it could make it more readable. It is very long and deep.